app/src/main/java/tech/relaycorp/gateway/domain/courier/GenerateCargo.kt

```diff
@@ -1,11 +1,69 @@
 package tech.relaycorp.gateway.domain.courier
 
+import tech.relaycorp.gateway.common.Logging.logger
+import tech.relaycorp.gateway.common.toZonedDateTime
+import tech.relaycorp.gateway.data.database.ParcelCollectionDao
+import tech.relaycorp.gateway.data.database.StoredParcelDao
+import tech.relaycorp.gateway.data.disk.DiskMessageOperations
+import tech.relaycorp.gateway.data.disk.MessageDataNotFoundException
+import tech.relaycorp.gateway.data.model.ParcelCollection
+import tech.relaycorp.gateway.data.model.RecipientLocation
+import tech.relaycorp.gateway.data.model.StoredParcel
+import tech.relaycorp.relaynet.cogrpc.readBytesAndClose
+import tech.relaycorp.relaynet.messages.ParcelCollectionAck
+import tech.relaycorp.relaynet.messages.payloads.CargoMessageWithExpiry
+import tech.relaycorp.relaynet.messages.payloads.batch
 import java.io.InputStream
+import java.util.logging.Level
 import javax.inject.Inject
 
 class GenerateCargo
-@Inject constructor() {
+@Inject constructor(
+    private val storedParcelDao: StoredParcelDao,
+    private val parcelCollectionDao: ParcelCollectionDao,
+    private val diskMessageOperations: DiskMessageOperations
+) {
 
-    // TODO: implementation
-    suspend fun generate(): Iterable<InputStream> = emptyList()
+    suspend fun generate(): Iterable<() -> InputStream> =
+        (getPCAsMessages() + getParcelMessages())
+            .asSequence()
+            .batch()
+            .map { { it.cargoMessageSet.serializePlaintext().inputStream() } }
+            .asIterable()
+
+    private suspend fun getPCAsMessages() =
+        parcelCollectionDao.getAll().map { it.toCargoMessageWithExpiry() }
+
+    private suspend fun getParcelMessages() =
+        storedParcelDao.listForRecipientLocation(RecipientLocation.ExternalGateway)
+            .mapNotNull { it.toCargoMessageWithExpiry() }
```

Comment:
When can it be `null`?

Reply:
When we can't read the Parcel from disk, in the method below. I'll make that method return type explicit, so it's clearer.